### PR TITLE
PRMT-3439-REFACTOR

### PR DESCRIPTION
**Pull request by martin-nhs**

### CODE REVIEW

File: src/test/java/uk/nhs/prm/e2etests/test/RepositoryE2ETest.java [hunk 1]
```diff
@@ -555,4 +586,26 @@ private void setManagingOrganisationToRepo(String nhsNumber) {
...
+    // ============ HELPER METHODS ============
+    private void addSmallEhrToEhrRepo(String nhsNumber) {
+        final String inboundConversationId = UUID.randomUUID().toString();
+        final String messageId = UUID.randomUUID().toString();
+        final SmallEhrTemplateContext smallEhrTemplateContext = SmallEhrTemplateContext.builder()
+                .inboundConversationId(inboundConversationId.toUpperCase())
+                .messageId(messageId.toUpperCase())
+                .nhsNumber(nhsNumber)
+                .build();
+        final String smallEhrMessage = this.templatingService.getTemplatedString(SMALL_EHR_WITHOUT_LINEBREAKS, smallEhrTemplateContext);
```
Comment (joefong-nhs) [inline src/test/java/uk/nhs/prm/e2etests/test/RepositoryE2ETest.java:599]:
```suggestion
        final SmallEhrTemplateContext smallEhrTemplateContext = SmallEhrTemplateContext.builder()
                .nhsNumber(nhsNumber)
                .build();
        final String inboundConversationId = smallEhrTemplateContext.getInboundConversationId().toLowerCase();
        final String smallEhrMessage = this.templatingService.getTemplatedString(SMALL_EHR_WITHOUT_LINEBREAKS, smallEhrTemplateContext);
```

I did some trial & error here, and it seems that we could let SmallEhrTemplateContext to generate the conversation Id. Just that the conversationId needs to be lowercased before passed to transferTrackerService.

...Or we could possibly modify TransferTrackerService and let it automatically lowercase any conversationId that it got from method call.

File: src/main/java/uk/nhs/prm/e2etests/queue/AbstractMessageQueue.java [hunk 2]
```diff
@@ -157,4 +161,32 @@ private boolean hasResolutionMessageNow(NemsResolutionMessage messageToCheck) {
...
+        if(expectedNumberOfMessages <= 0 || secondsToPoll <= 0)
+            throw new IllegalArgumentException("Expected number of messages or seconds to poll must be greater than 0.");
+
+        log.info("Attempting to get at least {} message(s) with substring {} on queue {}.", expectedNumberOfMessages, substring, this.queueUri);
+
+        List<SqsMessage> allMessages = new ArrayList<>();
+
+        try {
+            await().atMost(secondsToPoll, TimeUnit.SECONDS).with().pollInterval(100, TimeUnit.MILLISECONDS).untilAsserted(
+                    () -> {
```
Comment (MohammadIqbalAD-NHS) [inline src/main/java/uk/nhs/prm/e2etests/queue/AbstractMessageQueue.java:175]: Totally optional, I prefer these brackets to be placed at the end of line 174.

File: src/main/java/uk/nhs/prm/e2etests/repository/TransferTrackerDatabaseRepository.java [hunk 3]
```diff
@@ -30,12 +30,16 @@ public TransferTrackerDatabaseRepository(
     }
 
     public void save(TransferTrackerRecord transferTrackerDynamoDbEntry) {
+        // Update the conversation ID so that it can be saved into th Transfer Tracker DynamoDB.
```
Comment (MohammadIqbalAD-NHS) [inline src/main/java/uk/nhs/prm/e2etests/repository/TransferTrackerDatabaseRepository.java:33]: Typo.
"th" should be "the".

File: src/test/java/uk/nhs/prm/e2etests/test/RepositoryE2ETest.java [hunk 4]
```diff
@@ -373,25 +374,55 @@ void shouldTransferRepresentativeSizesAndTypesOfEhrs_DevOnly(Gp2GpSystem sourceS
                 TimeUnit.MINUTES))
                 .isNotNull();
 
-        assertTrue(transferTrackerService.isStatusForConversationIdPresent(triggerMessage.getConversationId(), TransferTrackerStatus.EHR_TRANSFER_TO_REPO_COMPLETE.status));
+        assertTrue(transferTrackerService.isStatusForConversationIdPresent(triggerMessage.getConversationId(), EHR_TRANSFER_TO_REPO_COMPLETE.status));
 
         // option: assert in ehr-repo - check all messages complete - evaluate need based on:
         //  - ehr out round trip testing
         //  - implementation of PRMT-2972
```
Comment (MohammadIqbalAD-NHS) [inline src/test/java/uk/nhs/prm/e2etests/test/RepositoryE2ETest.java:381]: This seems like a redundant comment in the middle of nowhere. Remove?

File: src/main/java/uk/nhs/prm/e2etests/queue/AbstractMessageQueue.java [hunk 5]
```diff
@@ -157,4 +161,32 @@ private boolean hasResolutionMessageNow(NemsResolutionMessage messageToCheck) {
...
+                        Optional<SqsMessage> found = findMessageContaining(substring);
+
+                        if (found.isPresent()) {
+                            boolean isNewMessage = allMessages.stream()
+                                    .noneMatch(sqsMessage -> sqsMessage.getId().equals(found.get().getId()));
+                            if (isNewMessage) allMessages.add(found.get());
+                        }
```
Comment (MohammadIqbalAD-NHS) [inline src/main/java/uk/nhs/prm/e2etests/queue/AbstractMessageQueue.java]: This looks sceptical..?

**Pull request merged**